Support some Webpack plugins

tnunamak opened the pull request:
The scoutfile build uses three Webpack plugins. They are DefinePlugin, UglifyJsPlugin and BannerPlugin. While it definitely doesn't make sense to try to resolve the use of two UglifyJsPlugin's, there are [many others](https://webpack.github.io/docs/list-of-plugins.html) that apps could use without getting in scoutfile's way.

This change compares the two plugin lists, checks the prototype names for any matches, and errors out if there are any (or if they can't be determined).

sirugh commented:
This looks good. Some notes:

* You chose not to add semicolons. Will that break future linting? Please run eslint to ensure conformity :)
* Your code is clear and readable, but has no comments explaining what exactly is going on. It may be beneficial to add some comments such as why you compare the constructor names, or what are you are doing when you `push` and then `delete` the old options.plugins array.

tnunamak commented:
> You chose not to add semicolons. Will that break future linting? Please run eslint to ensure conformity.

Good catch. ESLint runs on commit, but there's no semicolon rule. I added the ones I was missing.

I also added some comments. Thanks for the feedback.

sirugh commented:
👍

sirugh commented on lib/index.js:
```diff
@@ -131,8 +131,37 @@ module.exports.generate = function generateScout(options, callback) {
+          // the same plugin. If we get an error and can't tell, assume the
+          // worst.
+          return Object.getPrototypeOf(plugin).constructor.name ===
+            Object.getPrototypeOf(optionPlugin).constructor.name
```
`;`

reason-bv commented on lib/index.js:
```diff
@@ -131,8 +131,37 @@ module.exports.generate = function generateScout(options, callback) {
+            Object.getPrototypeOf(optionPlugin).constructor.name;
+        }
+        catch (e) {
+          throw new Error('Cannot verify that the plugin types included in ' +
```
From a style point of view, this is ugly, and should be

```
throw new Error(
  'Cannot verify that the plugin types included in options are allowed'
);
```

even if that exceeds 80 characters.

reason-bv commented on lib/index.js:
```diff
@@ -131,8 +131,37 @@ module.exports.generate = function generateScout(options, callback) {
+
+  // Copy user-specified plugins to our Webpack options, then delete
+  // them so the entire plugin list does not get overwritten.
+  Array.prototype.push.apply(webpackOptions.plugins,
```
Similarly, I'd prefer to see

```
 Array.prototype.push.apply(
  webpackOptions.plugins,
  options.webpackOptions.plugins
);
```

reason-bv commented:
I have stylistic complaints only. Otherwise 👍 .

tnunamak commented:
Thanks for the feedback. FYI, I had to append `// eslint-disable-line max-len` to overcome the pre-commit lint check.

sirugh commented:
👍 And maybe we can revisit the max length requirements at a future time.
